### ISSUE TITLE
Support Firefox for Android

### ISSUE THREAD

**Issue by jefft0** [ipfs/ipfs-companion#109]
A simplified Firefox for Android add-on only needs the preference for Custom Gateway Host. Then I can point it to my home IPFS node to open a URI like fs:/ipfs/Qm.... and browse the permanent web.

The simplified add-on doesn't need the IPFS Web UI since there is no local IPFS daemon.

**Comment by Kubuxu** [comment]
I think handling should go through this: https://github.com/ligi/IPFSDroid

For now it is a place holder but they aim at providing unified IPFS interface on Android

**Comment by jefft0** [comment]
I tried IPFSDroid. When I click an fs: URL in Firefox, it opens the external IPFSDroid application. I think a Firefox add-on is still necessary to rewrite the URL while staying in Firefox. Also, I'm only using the Firefox add-on to redirect to a Custom Gateway. The purpose of IPFSDroid is to run IPFS on the phone which is different. 

Is it possible to have a version of the add-on which only rewrites fs: to use a Custom Gateway?

**Comment by lidel** [comment]
The "problem" is that the addon already does it (if you disabled redirect).

For example:
> `fs:/ipfs/QmPQN7bUeCsYpLzkcNRWicE6C2o4XAG96x2AANNRz9J3aL` 
> redirects to 
> https://ipfs.io/ipfs/QmPQN7bUeCsYpLzkcNRWicE6C2o4XAG96x2AANNRz9J3aL

On desktop the addon does what you want, but on Android there is additional moving piece: [intents](https://developer.android.com/reference/android/content/Intent.html).

I think the issue is that Android prefers IPFSDroid for `fs:` because it [lists the protocol as a supported one](http://stackoverflow.com/questions/16372893/handling-custom-url-for-http-protocol-in-android) while Firefox does not.  

Not sure if it is possible for Firefox to dynamically extend its Android Manifest with protocols handled by 3rd-party addons.  

Requires further research.

**Comment by lidel** [comment]
_Short answer:_ we did not test/support add-on on Android yet, but fetching site assets from `fs:/ipfs/Qm..` is _technically_ possible. WebExtension version of this add-on will support it.

---

_Longer answer:_ I did some preliminary research regarding Firefox for Android and these my findings.
Forgive me going into low level technical details, this is also sort of a node-to-self.

- **It does not work right now.** I am unable to install [v1.5.9](https://github.com/lidel/ipfs-firefox-addon/releases/tag/v1.5.9) on Firefox for Android on my Nexus 5.
  According to page [Extensions for Firefox for Android](https://developer.mozilla.org/en-US/Add-ons/Firefox_for_Android) some of XUL internals / parts of old SDK (which we rely on right now) are [not supported on Mobile](https://developer.mozilla.org/en-US/Add-ons/SDK/Tutorials/Mobile_development#Module_Compatibility). This explains why add-on is hidden on AMO pages and I had to open the direct link to attempt install (which did not work).

- Good news: [WebExtensions](https://developer.mozilla.org/en-US/Add-ons/WebExtensions)  are supported since Firefox 48 for Android:

  > With the release of Firefox 48 we are announcing Android support for WebExtensions. WebExtensions add-ons can now be installed and run on Android, just like any other add-on. 

  with a caveat:

  > [..] because Firefox for Android makes use of a native user interface, anything that involves user interface interaction is currently unsupported (similar to existing extensions on Android).

In short, **Android support in on the Roadmap** as two steps:

1. Move Add-on to WebExtensions (https://github.com/lidel/ipfs-firefox-addon/issues/20)  
2. Ensure it works on mobile (add GUI workarounds, if needed)

**Comment by facastagnini** [comment]
It will be great to have this plugin working on firefox for android... let me know if you need a tester.

**Comment by lidel** [comment]
> With the release of Firefox 48 we are announcing Android support for WebExtensions. WebExtensions add-ons can now be installed and run on Android, just like any other add-on. However, because Firefox for Android makes use of a native user interface, anything that involves user interface interaction is currently unsupported (similar to existing extensions on Android).
– https://blog.mozilla.org/addons/2016/04/29/webextensions-in-firefox-48/

Added this issue to the scope of [v2.0.0](https://github.com/lidel/ipfs-firefox-addon/milestone/3).

**Comment by lidel** [comment]
Quick status update from Mozilla about Android goals for firefox 57:

> We aim to complete enough APIs to allow the most popular add-ons on Android (AdBlock Plus and uBlock Origin) to run as WebExtensions. To support those we will complete: browserAction, pageAction, options_ui, permissions and maybe contextMenus (being investigated at this time).
> 
> There is no plan to support other APIs like history, bookmarks etc by Firefox 57.

**Comment by lidel** [comment]
Another [update](https://blog.mozilla.org/addons/2017/06/14/webextensions-firefox-55/) from Mozilla:

> Firefox 55 for Android has gained browserAction support. Currently a textual menu item is added to the bottom of the menu on Android. It supports browserAction.onClicked and setTitle and getTitle. Tabs support was added to pageAction.

**Comment by lidel** [comment]
Quick status update with tickets that are blocking this effort:
- [Support `web-ext run` on Firefox for Android (fennec) #737](https://github.com/mozilla/web-ext/issues/737) (OPEN)
- [Bug 1185785 - [tracking] Extension API for Android](https://bugzilla.mozilla.org/show_bug.cgi?id=1185785) (OPEN, meta-ticket for Android support)
    - [Bug 1302504 - Add support for options_ui on Android](https://bugzilla.mozilla.org/show_bug.cgi?id=1302504) ( FIXED in Firefox 56)
    - [Bug 1364945 - [runtime] runtime.openOptionsPage does not work on Android](https://bugzilla.mozilla.org/show_bug.cgi?id=1364945) (OPEN)
    - [Bug 1330159 - [tracking] BrowserAction API support for Android WebExtensions](https://bugzilla.mozilla.org/show_bug.cgi?id=1330159) (OPEN)

**Comment by jefft0** [comment]
Should IPFS Companion 2.0.13 work in Firefox for Android? I can install it, but I can't open the Preferences to change the Custom Gateway.

**Comment by lidel** [comment]
Hm.. indeed, opening Preferences was implemented in Firefox 57, which is the current Beta (will be released as stable in November).
I changed minimal version at AMO to Firefox 57.

If you want to test it right now, install [Firefox for Android Beta](https://play.google.com/store/apps/details?id=org.mozilla.firefox_beta&hl=en) and give it a try.

**Comment by jefft0** [comment]
Thanks! Firefox Beta works great. The future has arrived!

**Comment by lidel** [comment]
@jefft0  I fixed CSS so that you dont need to zoom/pan on Android anymore 👌 
If you want to test it, opt-in for [2.0.15beta1](https://addons.mozilla.org/en-US/firefox/addon/ipfs-companion/versions/2.0.15beta1)
